web/data/accounts.go

```diff
@@ -0,0 +1,218 @@
+package data
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"net/url"
+
+	"github.com/cozy/cozy-stack/pkg/accounts"
+	"github.com/cozy/cozy-stack/pkg/consts"
+	"github.com/cozy/cozy-stack/pkg/couchdb"
+	perms "github.com/cozy/cozy-stack/pkg/permissions"
+	"github.com/cozy/cozy-stack/web/jsonapi"
+	"github.com/cozy/cozy-stack/web/middlewares"
+	"github.com/cozy/cozy-stack/web/permissions"
+	"github.com/labstack/echo"
+)
+
+// TODO: make specific routes for managing accounts. The overriding of the
+// /data/io.cozy.accounts/* routes is here mainly for retro-compatible reasons,
+// but specific routes would improve the API.
+
+func accountDoctype(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		doctype := consts.Accounts
+		c.Set("doctype", doctype)
+		docidraw := c.Param("docid")
+		docid, err := url.QueryUnescape(docidraw)
+		if err != nil {
+			return jsonapi.NewError(http.StatusBadRequest, "Invalid docid '%s'", docid)
+		}
+		c.Set("docid", docid)
+		return next(c)
+	}
+}
+
+func getAccount(c echo.Context) error {
+	instance := middlewares.GetInstance(c)
+	doctype := consts.Accounts
+	docid := c.Get("docid").(string)
+	if docid == "" {
+		return dbStatus(c)
+	}
+
+	var out couchdb.JSONDoc
+	err := couchdb.GetDoc(instance, doctype, docid, &out)
+	if err != nil {
+		return fixErrorNoDatabaseIsWrongDoctype(err)
+	}
+	out.Type = doctype
+
+	if err = permissions.Allow(c, permissions.GET, &out); err != nil {
+		return err
+	}
+
+	if encryptAccount(out) {
+		if err = couchdb.UpdateDoc(instance, out); err != nil {
+			return err
+		}
+	}
+
+	perm, err := permissions.GetPermission(c)
```

Comment:
I'd prefer if we check the permissions before doing the `encryptAccount`. No good reason, it's just that is seems more logical.

Reply:
The permissions are already checked before, line 52. This is a specific check for the token type.